### DISCUSSION TITLE
Maven Parent lookup error

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#25758] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

No lookup errors in the repo

### What you tried so far.

https://github.com/rarkins/WebGoatNorman/issues/3

This repo has lookup warnings for `org.owasp.webgoat.lesson:webgoat-lessons-parent`:

![image](https://github.com/renovatebot/renovate/assets/6311784/a9c8cba3-5ac7-46fe-9744-f8817a5f7a9d)

Is it a mistake that Renovate is attempting to lookup this package from Maven central, or is something else going on?

@Churro would you mind advising? 

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by Churro** [comment]
renovate shouldn't lookup this package on Maven Central, as it is defined in the local repository [here](https://github.com/rarkins/WebGoatNorman/blob/develop/webgoat-lessons/pom.xml).

The current behavior is an effect of PR https://github.com/renovatebot/renovate/pull/18788:
The objective was to mark dependencies with `depType = "parent-root"` if a dependency was defined in a local parent `pom.xml`. A renovate workaround would then disable updates for them. 

**In this case, `depType = "parent-root"` is not set and therefore the dependency is not skipped, although it should have been.**

What happens:
- renovate extracts `org.owasp.webgoat.lesson:webgoat-lessons-parent` e.g., from [here ](https://github.com/rarkins/WebGoatNorman/blob/develop/webgoat-lessons/xxe/pom.xml#L6-L10)and marks it with `depType = "parent"` https://github.com/renovatebot/renovate/blob/d7e741a184dfcdb2bca0ebcea7ad09ad6333c13f/lib/modules/manager/maven/extract.ts#L75-L76
- renovate resolves the parent pom.xml https://github.com/renovatebot/renovate/blob/d7e741a184dfcdb2bca0ebcea7ad09ad6333c13f/lib/modules/manager/maven/extract.ts#L308-L312
- renovate adds parent dependencies to `rootDeps` **if the containing pom.xml itself has no local parent pom.xml**
https://github.com/renovatebot/renovate/blob/d7e741a184dfcdb2bca0ebcea7ad09ad6333c13f/lib/modules/manager/maven/extract.ts#L436-L441
- Matching dependencies are eventually marked with `depType = "parent-root"` https://github.com/renovatebot/renovate/blob/d7e741a184dfcdb2bca0ebcea7ad09ad6333c13f/lib/modules/manager/maven/extract.ts#L451-L453

What prevents `parent-root` from being set in this case is the fact that `webgoat-lessons/pom.xml` also has a local parent pom.xml. This causes the second clause of this `if` to fail:
https://github.com/renovatebot/renovate/blob/d7e741a184dfcdb2bca0ebcea7ad09ad6333c13f/lib/modules/manager/maven/extract.ts#L438

**Solution:** Remove the second if-clause. Existing tests still pass **but** there could be side effects I'm not aware of. I assume there was some intention to include this check in the first place, so probably it could be clarified first if it's safe to be removed.

**Reply by rarkins** [reply]
Thanks, @Churro !

@zharinov you added this code as part of https://github.com/renovatebot/renovate/pull/18788

Can you take a look?

**Reply by SebastianGoeb** [reply]
Is there any progress on this issue? I just stumbled upon this issue (renovate tries to update a locally defined parent, which is nonsense of course).

If I understand correctly, local parent pom.xmls should be detected as `parent-root` but are instead detected as "only" `parent`. This causes them to incorrectly NOT be skipped by the preset parent-root exclusion. It does only cause some unsightly warning logs, but it would be nice to get the parent-root detection fixed correctly.

Although I must say: a local maven parent doesn't necessarily have to be the root pom.xml. I can just as well be a sibling maven module, like the following. Is `parent-root` really the best term for a "parent that shouldn't be updated because it is locally defined"?

- pom.xml (not parent pom)
- module-my-parent (this is the parent pom for the other modules)
- module-my-app1 (`<parent>module-my-parent</parent>`)
- module-my-app2 (`<parent>module-my-parent</parent>`)

**Reply by Churro** [reply]
@SebastianGoeb , I think the `parent-root` term was mainly needed as an identifier for this preset: https://docs.renovatebot.com/presets-workarounds/#workaroundsdisablemavenparentroot

Generally, this issue is still unresolved. Probably it would make sense to revise / re-implement the handling of local parent packages in the maven manager with a different approach.

**Reply by SebastianGoeb** [reply]
I couldn't quite figure out what that workaround is for either, if not for this exact issue. Why would one want to skip updates to maven packages that are project root. Isn't that just the exact issue being discussed here, minus the uncommon edge case I described?

**Reply by Bifi1992** [reply]
## renovate version
```
$ renovate -v
37.440.7
```

## answer to @SebastianGoeb 
> Why would one want to skip updates to maven packages that are project root.

in my case, I usually only check-in `0.0.1-SNAPSHOT` versions into my repositories. Unfortunately, this leads to renovate PRs for every module in the project once I've published a new artifact version.

Meaning, in the following simple project structure, renovate will always try to open a PR for the `pom.xml` of the modules when a new version of the root `pom.xml` is published:
```
|
|_ projectRoot/
   |_ moduleA/
   |  |_ pom.xml  # <-- "prTitle: \"Update dependency my.group.id:projectRootParent to vX.Y.Z\""
   |_ moduleB/
   |  |_ pom.xml  # <-- "prTitle: \"Update dependency my.group.id:projectRootParent to vX.Y.Z\""
   |_ pom.xml 
```

## my current workaround
Due to this, I've currently configured renovate to **disable all Maven parent upgrades** like this:
```
{
  "packageRules": [
    {
      "matchDatasources": [ "maven" ],
      "matchPackagePrefixes": [
        "my.group.id"
      ],
      "matchDepTypes": [ "parent" ],
      "enabled": false
    }
  ]
}
```
Which is sad, as **I would very much like to be able to automatically update the parent of my root pom, but not the parent version of the modules**.

## Trial with [workarounds:disableMavenParentRoot](https://docs.renovatebot.com/presets-workarounds/#workaroundsdisablemavenparentroot)
* I've removed the previous `packageRule` that disabled all parent version updates
* I've added the workaround's `packageRule`
```
{
  "packageRules": [
    {
      "enabled": false,
      "matchDepTypes": [
        "parent-root"
      ],
      "matchManagers": [
        "maven"
      ]
    }
  ]
}
```
* No effect. The renovate run still tries to update the parent version within the modules `pom.xml` files.

Am I doing something planely wrong besides checking in `0.0.1-SNAPSHOT` versions? This feels like a basic use case for Maven projects, right?


EDIT: I guess, your question still prevails, @SebastianGoeb. The problematic PRs are in the modules, not the root.

**Reply by karouf** [reply]
Like @Djiit said, we are encountering the same issue and just opened [a PR](https://github.com/renovatebot/renovate/pull/34533) with a possible fix to that.

**Comment by Djiit** [comment]
Hello team, we're also encounrering this on spring-boot based repositories. Basically, a local dependency (the "***-parent" thing) is considered as to be fetched somewhere. Any known workaround in 2025?
